File: app/src/main/java/me/zhanghai/android/files/glide/MediaEmbeddedPictureModelLoader.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.glide;

import android.media.MediaMetadataRetriever;
import android.os.ParcelFileDescriptor;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.load.data.mediastore.MediaStoreUtil;
import com.bumptech.glide.load.model.ModelLoader;
import com.bumptech.glide.load.model.ModelLoaderFactory;
import com.bumptech.glide.load.model.MultiModelLoaderFactory;
import com.bumptech.glide.signature.ObjectKey;

import java.nio.ByteBuffer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.Path;
import me.zhanghai.android.files.file.MimeTypes;
import me.zhanghai.android.files.provider.document.DocumentFileSystemProvider;
import me.zhanghai.android.files.provider.document.resolver.DocumentResolver;
import me.zhanghai.android.files.provider.linux.LinuxFileSystemProvider;

public class MediaEmbeddedPictureModelLoader implements ModelLoader<Path, ByteBuffer> {

    @Override
    public boolean handles(@NonNull Path model) {
        if (!(LinuxFileSystemProvider.isLinuxPath(model)
                || DocumentFileSystemProvider.isDocumentPath(model))) {
            return false;
        }
        Path fileName = model.getFileName();
        if (fileName == null) {
            return false;
        }
        String mimeType = MimeTypes.getMimeType(fileName.toString());
        return MimeTypes.isMedia(mimeType);
    }

    @Nullable
    @Override
    public LoadData<ByteBuffer> buildLoadData(@NonNull Path model, int width, int height,
                                              @NonNull Options options) {
        if (MediaStoreUtil.isThumbnailSize(width, height) && !GlidePathUtils.shouldLoadThumbnail(
                model)) {
            return null;
        }
        return new LoadData<>(new ObjectKey(model), new Fetcher(model));
    }

    private static class Fetcher implements DataFetcher<ByteBuffer> {

        @NonNull
        private final Path path;

        public Fetcher(@NonNull Path path) {
            this.path = path;
        }

        @Override
        public void loadData(@NonNull Priority priority,
                             @NonNull DataCallback<? super ByteBuffer> callback) {
            ByteBuffer picture;
            try {
                MediaMetadataRetriever retriever = new MediaMetadataRetriever();
                try {
                    setDataSource(retriever, path);
                    picture = ByteBuffer.wrap(retriever.getEmbeddedPicture());
                } finally {
                    retriever.release();
                }
            } catch (Exception e) {
                callback.onLoadFailed(e);
                return;
            }
            callback.onDataReady(picture);
        }

        private static void setDataSource(@NonNull MediaMetadataRetriever retriever,
                                          @NonNull Path path) throws Exception {
            if (LinuxFileSystemProvider.isLinuxPath(path)) {
                retriever.setDataSource(path.toFile().toString());
            } else if (DocumentFileSystemProvider.isDocumentPath(path)) {
                try (ParcelFileDescriptor pfd = DocumentResolver.openParcelFileDescriptor(
                        (DocumentResolver.Path) path, "r")) {
                    retriever.setDataSource(pfd.getFileDescriptor());
                }
            } else {
                throw new AssertionError(path);
            }
        }

        @Override
        public void cleanup() {}

        @Override
        public void cancel() {}

        @NonNull
        @Override
        public Class<ByteBuffer> getDataClass() {
            return ByteBuffer.class;
        }

        @NonNull
        @Override
        public DataSource getDataSource() {
            return DataSource.LOCAL;
        }
    }

    public static class Factory implements ModelLoaderFactory<Path, ByteBuffer> {

        @NonNull
        @Override
        public ModelLoader<Path, ByteBuffer> build(@NonNull MultiModelLoaderFactory multiFactory) {
            return new MediaEmbeddedPictureModelLoader();
        }

        @Override
        public void teardown() {}
    }
}

File: app/src/main/java/me/zhanghai/android/files/glide/MediaEmbeddedPictureModelLoader.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.glide

import android.media.MediaMetadataRetriever
import com.bumptech.glide.Priority
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.load.data.mediastore.MediaStoreUtil
import com.bumptech.glide.load.model.ModelLoader
import com.bumptech.glide.load.model.ModelLoader.LoadData
import com.bumptech.glide.load.model.ModelLoaderFactory
import com.bumptech.glide.load.model.MultiModelLoaderFactory
import com.bumptech.glide.signature.ObjectKey
import java8.nio.file.Path
import me.zhanghai.android.files.file.MimeType
import me.zhanghai.android.files.file.guessFromPath
import me.zhanghai.android.files.file.isMedia
import me.zhanghai.android.files.provider.document.isDocumentPath
import me.zhanghai.android.files.provider.document.resolver.DocumentResolver
import me.zhanghai.android.files.provider.linux.isLinuxPath
import java.nio.ByteBuffer

class MediaEmbeddedPictureModelLoader : ModelLoader<Path, ByteBuffer> {
    override fun handles(model: Path): Boolean {
        if (!(model.isLinuxPath || model.isDocumentPath)) {
            return false
        }
        val fileName = model.fileName ?: return false
        return MimeType.guessFromPath(fileName.toString()).isMedia
    }

    override fun buildLoadData(
        model: Path, width: Int, height: Int,
        options: Options
    ): LoadData<ByteBuffer>? {
        if (MediaStoreUtil.isThumbnailSize(width, height) && !model.shouldLoadThumbnail) {
            return null
        }
        return LoadData(ObjectKey(model), Fetcher(model))
    }

    private class Fetcher(private val path: Path) : DataFetcher<ByteBuffer> {
        override fun loadData(
            priority: Priority,
            callback: DataFetcher.DataCallback<in ByteBuffer>
        ) {
            val picture = try {
                val retriever = MediaMetadataRetriever()
                try {
                    setDataSource(retriever, path)
                    ByteBuffer.wrap(retriever.embeddedPicture)
                } finally {
                    retriever.release()
                }
            } catch (e: Exception) {
                callback.onLoadFailed(e)
                return
            }
            callback.onDataReady(picture)
        }

        override fun cleanup() {}

        override fun cancel() {}

        override fun getDataClass(): Class<ByteBuffer> = ByteBuffer::class.java

        override fun getDataSource(): DataSource = DataSource.LOCAL

        companion object {
            @Throws(Exception::class)
            private fun setDataSource(retriever: MediaMetadataRetriever, path: Path) {
                when {
                    path.isLinuxPath -> retriever.setDataSource(path.toFile().toString())
                    path.isDocumentPath ->
                        DocumentResolver.openParcelFileDescriptor(
                            (path as DocumentResolver.Path), "r"
                        ).use { pfd -> retriever.setDataSource(pfd.fileDescriptor) }
                    else -> throw AssertionError(path)
                }
            }
        }
    }

    class Factory : ModelLoaderFactory<Path, ByteBuffer> {
        override fun build(multiFactory: MultiModelLoaderFactory): ModelLoader<Path, ByteBuffer> =
            MediaEmbeddedPictureModelLoader()

        override fun teardown() {}
    }
}
